Clarify integration and API test distinction (#11)
ExtensionEngine/tech-guide: recipes/automated-testing.md

### Integration Tests

> With these tests, we test the combined behavior of multiple components within the system.

Reviewer: This phrasing might be more intuitive (behaving together part, compared to combined behavior).
I might be subjective and wrong, so feel free to resolve if it doesn't make sense. :)
```suggestion
With these tests, we test how multiple components of the system behave together.
```

#### Infrastructure

Running the tests on test infrastructure should be preferred to mocking, unlike in unit tests. Ideally, a full application instance would be run, to mimic real application behavior as close as possible.
This usually includes running the application connected to a test database, inserting fake data into it during the test setup, and doing assertions on the current state of the database. This also means integration test code should have full access to the test infrastructure for querying. 
> [!NOTE]

> > Regardless of whether using raw queries or the ORM, simple queries should be used to avoid introducing business logic within tests.

Reviewer: @dbettini I don't understand what's meant by this. Can you elaborate a bit, please (in reply to this comment)?

Author: I'm guessing you're asking about the note about simple queries? If yes, I'm talking about using simple SELECTs with WHERE, instead of using more complex custom queries, since that could result in the test giving a false positive/negative. E.g. query being too specific, making it impossible to realize that something else exists in the database which would fail the test.

Reviewer 2: I understood it as "don't repeat business logic in tests". So using simple queries to check if something is changed (or unchanged) is preferred over repeating a business model query and asserting the output is correct.

Author: @MiroDojkic queries used to fetch the data that will be asserted on

However, mocking can still be used when needed, for example when expecting side-effects that call third party services.

#### Entry points

Integration test entry points can vary depending on the application use cases. These include services, controllers, or the API. These are not set in stone and should be taken into account when making a decision. For example:
- A use case that can be invoked through multiple different protocols can be tested separately from them, to avoid duplication. A tradeoff in this case is the need to write some basic tests for each of the protocols.

> - A use case that will always be invokeable through a single protocol might benefit enough from only being tested using that protocol. E.g. a HTTP API route test might eliminate the need for a lower level, controller/service level test. This would also enable the testing of authorization within the same test, which might not have been possible otherwise depending on the technology used.

Reviewer: Authorization should be tested separately to increase determinism (if authorization fails, we know it failed immediately).
I assume what you meant here is that this way, we test the integration of the auth layer with other app layers?

Author: Yes, was referring to being able to expect a 403 if trying to access an admin route with a non-admin token, for example. Maybe replace `This would also enable the testing of authorization within the same test, which might not have been possible otherwise depending on the technology used.` with `This would also enable the inclusion of the auth layer within these tests, which might not have been possible otherwise depending on the technology used.`?

Reviewer 3: @dbettini That makes more sense to me. Otherwise it might be interpreted as "auth is implicitly tested through other integration tests" so making it clear that this covers testing auth integration makes sense.

> referring to being able to expect a 403 if trying to access an admin route 

Regarding ⬆️ , I guess there are 2 test-cases I see here:
1. Testing is our auth layer working as expected i.e. if for given auth permissions appropriate access is granted
2. Are permissions set properly on specific route/entry point
I think the first should be tested generically in a auth-specific tests while the second we can get implicitly from integration test on that specific entry point. While I see value in testing appropriate permissioning on a entry point, I'm wondering is it worth-while to add (all) auth permutation assertions to each entry point as, depending on the complexity of auth layer and permissioning implemented, it could blow-up number of test-cases. 
What do you think? (I might be looking too much into a simple example 😄 )

Author: @kronicker unfortunately, I don't have a catch-all answer there 😄 I do like to test the basic cases, e.g. a route only being accessible to admins, for two reasons: 
1. Other than implicitly testing the behavior of the auth layer, it also tests that the correct rules were applied. This is useful because even if our auth layer works correctly, it still doesn't mean it was correctly configured for this specific route.
2. The level of danger in the case of an error - I feel much safer if a route that e.g. allows destructive actions for admins has a test that ensures other roles would get an error

But like you said, we should make sure to avoid combinatorial explosions, so it needs to be applied on a case by case basis IMO - this is why I didn't go into too much detail in this section, but went deeper in the `When not to use` section.

> Multiple approaches can be used within the same application depending on the requirements, to provide sufficient coverage.

Reviewer: Do we need to define a "use-case" to ensure different readers don't come up with different understandings?

Author: Hm, to be honest none comes to mind that wouldn't fit here. Do you have anything specific in mind?

Reviewer: I think for a junior, "use-case" can mean method invocation, while in software architecture, use-case usually relates to business logic use-case encapsulated in an app service method, command or query (in CQRS).

Author: Never heard of it being used that way, maybe because it was explained as a business logic use case even in college in my case, which might not be the case for everyone. 
Do you think rewording "application use cases" to "business logic use cases" is enough, or you have something else in mind?

Author: Added to glossary, backlinks seemed like overkill for now

- To verify the API endpoint performs authentication and authorization.
- To verify user permissions for that endpoint.
- To verify that invalid input is correctly handled.
- To verify the basic business logic is handled correctly, both in expected success and failure cases.
- To verify infrastructure related side-effects, e.g. database changes or calls to third party services.

#### When **not** to use

> - For extensive testing of business logic permutations beyond fundamental scenarios. Integration tests contain more overhead to write compared to unit tests and can easily lead to a combinatorial explosion. Instead, unit tests should be used for thorough coverage of these permutations.

Reviewer: What would combinatorial explosion mean, and how would it look like?

Author: It is the fast growth in the number of combinations that need to be tested when multiple business rules are involved. At least that's the term I've seen sometimes used in the context of writing tests 😄 It's especially important in integration tests since they have a higher chance of invoking multiple different business rules at the same time. Didn't explain the term because I guessed it could be googled. What do you think? Should a different term be used?

Reviewer 2: It falls in line with what was previously written that unit tests should be preferred way of testing each unit logic instead of testing multiple units and trying to add enough permutations/combinations to cover all use cases.
